### DISCUSSION TITLE
Problems related to using Elasticsearch in orchardCMS

### DISCUSSION THREAD

**Discussion by kharaei** [OrchardCMS/OrchardCore#16423] [Q&A]
When I activate the ElasticSearch feature according to the instructional [video](https://www.youtube.com/watch?v=7Mx1Vjsy3Xw&t=246s) 
1- The Elasticsearch menu does not get added under the Search submenu.
2- In the search -> settings -> search menu, there is no option to select the default search provider, and I only see a save button.
3- I have checked the enabled options in features and set them exactly as in the instructional video. However, the /search page and the content definitions -> content types menu encounter the following error: 
`InvalidOperationException: Unable to resolve service for type 'OrchardCore.Search.Elasticsearch.Core.Services.ElasticIndexSettingsService' while attempting to activate 'OrchardCore.Search.Elasticsearch.Drivers.ContentPickerFieldElasticEditorSettingsDriver'.`

Thank you for your guidance.

**Comment by Piedone** [comment]
I assume you run 1.8.x.. In the latest `main` this is explicitly shown on the UI, but you have to first configure Elasticsearch settings, as documented [here](https://docs.orchardcore.net/en/latest/reference/modules/Elasticsearch/#elasticsearch-configuration).

**Reply by kharaei** [reply]
I am using version 1.8.3 and have applied configurations according to the documentation:
1. I created Elasticsearch and Kibana using the same yml file provided within the documentation.
2. My appsettings.json file looks like this:
```json
{
  "Logging": {
    "IncludeScopes": false,
    "LogLevel": {
      "Default": "Warning",
      "Microsoft.Hosting.Lifetime": "Information"
    }
  },
  "OrchardCore_Elasticsearch": {
    "ConnectionType": "SingleNodeConnectionPool",
    "Url": "http://localhost",
    "Ports": [ 9200 ],
    "IndexPrefix": "orchardcore_elasticsearch_index",
    "Analyzers": {
      "standard": {
        "type": "standard"
      }
    }
  }
}
```
3. The following features have been enabled:
* Search
* Indexing
* Elasticsearch
* Elasticsearch Content Picker

According to the video, the Elasticsearch option under the search->settings menu should be enabled in this configuration. However, this is not happening.
Additionally, when I search for text in the search box (on the main page), I encounter the following error:
"Unable to resolve service for type 'OrchardCore.Search.Elasticsearch.Core.Services.ElasticIndexManager' while attempting to activate 'OrchardCore.Search.Elasticsearch.Services.ElasticsearchService'."

Furthermore, I have added the following package to the project, but it did not provide any benefit:
OrchardCore.Search.Elasticsearch.Core

**Reply by Piedone** [reply]
What do you see in the logs?

**Reply by kharaei** [reply]
> What do you see in the logs?

When I run the project, right at the start, without throwing an error, it logs this message:

```
OrchardCore.Search.Elasticsearch.Startup: Error: Elasticsearch is enabled but not active because the configuration is missing.
fail: OrchardCore.Search.Elasticsearch.Startup[0]
      Elasticsearch is enabled but not active because the configuration is missing.
```

**Reply by Piedone** [reply]
Your problem is "Elasticsearch is enabled but not active because the configuration is missing.", so as mentioned, you have to configure it fully. Note that your configuration is missing a `Username` and `Password`, what in the case of a local deployment is just "admin" for both.

You mustn't register `ElasticIndexManager` yourself. It'll be registered if you correctly configure the feature.

**Comment by kharaei** [comment]
I added the Elastic service in the `program.cs` file:
```csharp
builder.Services.AddOrchardCms().ConfigureServices(options =>
{
    var settings = new ConnectionSettings(new Uri("http://localhost:9200")).DefaultIndex("orchardcore_elasticsearch_index");
    var client = new ElasticClient(settings);
    options.AddSingleton<IElasticClient>(client);
    options.AddElasticServices();
});
```

it worked!

**Comment by kharaei** [comment]
Thank you. 
I finally found the problem. The issue I had was that I defined the configuration keys (inside appsettings) as OrchardMS instead of OrchardCore! 😅 Therefore, it couldn't read them, and the menus for Elasticsearch weren't active.
